### PR TITLE
better query param handling

### PR DESCRIPTION
**Pull request by czardoz**

### CODE REVIEW

File: lib/collection/query-param.js [hunk 1]
```diff
@@ -121,29 +138,31 @@ _.assign(QueryParam, /** @lends QueryParam */ {
...
-            }, []);
-        }
+        var encode = options && options.encode;
+
+        return params ? params.map(function (param) {
+            var key = param.key,
+                value = param.value;
+
+            if (value === undefined) {
+                return '';
```
Comment (kunagpal) [inline lib/collection/query-param.js]: Can be replaced with `EMPTY`

File: lib/collection/query-param.js [hunk 2]
```diff
@@ -121,29 +138,31 @@ _.assign(QueryParam, /** @lends QueryParam */ {
...
-            accumulator.length && (accumulator += '&');
-            accumulator += encodeURI(QueryParam.unparseSingle(param));
-            return accumulator;
-        }, '') : _.reduce(params, function (accumulator, param) {
-            accumulator.length && (accumulator += '&');
-            accumulator += QueryParam.unparseSingle(param);
-            return accumulator;
-        }, '');
-
-        // Special handling to ensure "{{" and "}}" are not URL encoded in the final query-string.
```
Comment (shamasis) [inline lib/collection/query-param.js]: bring me back

File: lib/collection/query-param.js [hunk 3]
```diff
@@ -121,29 +138,31 @@ _.assign(QueryParam, /** @lends QueryParam */ {
      * @returns {string}
      */
     unparse: function (params, options) {
-        var qs,
-            encode = options && options.encode;
-
-        if (!_.isArray(params)) {
```
Comment (shamasis) [inline lib/collection/query-param.js]: accept me

File: lib/collection/query-param.js [hunk 4]
```diff
@@ -153,10 +172,10 @@ _.assign(QueryParam, /** @lends QueryParam */ {
      * @returns {String}
      */
     unparseSingle: function (obj) {
-        if (!obj) { return ''; }
-        var unparsed = obj.key || '';
+        if (!obj) { return EMPTY; }
+        var unparsed = obj.key || EMPTY;
         // if the value is not null, then we add the value
-        !_.isNil(obj.value) && (unparsed += ('=' + (obj.value || '')));
+        !_.isNil(obj.value) && (unparsed += ('=' + (obj.value || EMPTY)));
```
Comment (shamasis) [inline lib/collection/query-param.js]: kill me `=`

File: test/unit/query-param.test.js [hunk 5]
```diff
@@ -32,4 +32,303 @@ describe('QueryParam', function () {
             paramStr = QueryParam.unparse(params, { encode: true });
         expect(paramStr).to.eql('x=y%25z');
     });
+
+    describe('parsing', function () {
```
Comment (shamasis) [inline test/unit/query-param.test.js]: can we add integrity tests?

**Comment by shamasis** [comment]
When you address the comment we put in that PR, can you first add a failing test case for them?

1. Non array structure 
2. Encoded {{}} character replacement.

File: lib/collection/query-param.js [hunk 6]
```diff
@@ -115,49 +133,63 @@ _.assign(QueryParam, /** @lends QueryParam */ {
...
-        }, '');
-
-        // Special handling to ensure "{{" and "}}" are not URL encoded in the final query-string.
-        (options && options.encode) && (qs = qs.replace(/%7B%7B/g, '{{')) && (qs = qs.replace(/%7D%7D/g, '}}'));
-        return qs;
+        str = params.map(function (param) {
+            return QueryParam.unparseSingle(param, encode);
+        }).join(AMPERSAND);
+
+        (encode) && (str = str.replace(/%7B%7B/g, '{{')) && (str = str.replace(/%7D%7D/g, '}}'));
```
Comment (kunagpal) [inline lib/collection/query-param.js]: Could be referenced as constants.

File: lib/collection/query-param.js [hunk 7]
```diff
@@ -43,13 +51,13 @@ _.assign(QueryParam.prototype, /** @lends QueryParam.prototype */ {
      */
     update: function (param) {
         _.assign(this, /** @lends QueryParam.prototype */ _.isString(param) ? QueryParam.parseSingle(param) : {
-            key: _.get(param, 'key') || '', // we do not replace falsey with blank string since null has a meaning
+            key: _.get(param, 'key') || EMPTY, // we do not replace falsey with blank string since null has a meaning
```
Comment (kunagpal) [inline lib/collection/query-param.js]: The code contradicts the comment.

**Pull request merged**